shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -399,136 +433,447 @@ class Context {
+      });
+    });
+  }
+
+  activateToken(tokenId) {
+    // Includes a new token (which must already be in `this.tokensById`) in our computation. The
+    // tricky part here is dealing with our already-accumulated knowledge; we need to compute how
+    // many of the token's requirements are currently unmet and whether we need to push anything new
+    // onto `setToTrueStack`.
+
+    check(tokenId, String);
+    if (tokenId in this.activeTokens) {
+      return false;
+    }
+
+    const token = this.tokensById[tokenId];
+    const grainId = token.grainId;
+    const viewInfo = this.grains[grainId].cachedViewInfo || {};
+    const tokenPermissions = PermissionSet.fromRoleAssignment(token.roleAssignment, viewInfo);
+
+    const sharerId = token.parentToken ? "t:" + token.parentToken : "i:" + token.identityId;
+    const recipientId = vertexIdOfTokenOwner(token);
+
+    let numUnmetRequirements = 0;
+    if (token.requirements) {
+      token.requirements.forEach((requirement) => {
+        if (requirement.permissionsHeld) {
+          const reqGrainId = requirement.permissionsHeld.grainId;
+          const reqVertexId = "i:" + requirement.permissionsHeld.identityId;
+          const reqPermissions = requirement.permissionsHeld.permissions || [];
+
+          forEachPermission(reqPermissions, (permissionId) => {
+            const variable = this.getVariable(reqGrainId, reqVertexId, permissionId);
+            if (!variable.value) {
+              numUnmetRequirements += 1;
+              variable.requirementsTailList.push(tokenId);
+            }
+          });
+        }
+      });
+    }
+
+    const activeToken = new ActiveToken(numUnmetRequirements, tokenPermissions.array);
+    let needToExploreUnmet = false;
+
+    activeToken.forEachPermission((permissionId, status) => {
+      if (!("doesNotCarry" in status)) {
+        const recipientVariable = this.getVariable(grainId, recipientId, permissionId);
+        if (!recipientVariable.value) {
+          const sharerVariable = this.getVariable(grainId, sharerId, permissionId);
+          if (!sharerVariable.value) {
+            sharerVariable.directTailList.push(tokenId);
+          } else {
+            activeToken.decrementDirect(permissionId);
+            if (numUnmetRequirements == 0) {
+              this.setToTrueStack.push({ grainId: grainId,
+                                         vertexId: vertexIdOfTokenOwner(token),
+                                         permissionId: permissionId,
+                                         responsibleTokenId: tokenId, });
+            } else {
+              needToExploreUnmet = true;
+            }
+          }
+        }
+      }
+    });
+
+    if (needToExploreUnmet) {
+      this.unmetRequirements.addRequirements(token.requirements);
+    }
+
+    this.activeTokens[tokenId] = activeToken;
+
+    return true;
+  }
+
+  _ensureVariableExists(grainId, vertexId, permissionId) {
```

Comment:
I think you can collapse _ensureVariableExists() into getVariable() at this point and save some redundant operations. Callers can call getVariable() and ignore the result if they just wanted to make sure the variable exists...